mlir/lib/Pass/IRPrinting.cpp

```diff
@@ -345,6 +346,67 @@ struct FileTreeIRPrinterConfig : public PassManager::IRPrinterConfig {
   llvm::DenseMap<Operation *, unsigned> counters;
 };
 
+/// A pass instrumenation to dump the IR before each pass into
+/// numbered files.
+/// It includes a mlir_reproducer info to rerun the pass.
+class ReproducerBeforeAll : public PassInstrumentation {
+public:
+  ReproducerBeforeAll(mlir::StringRef outputDir) : outputDir(outputDir) {}
+  void runBeforePass(Pass *pass, Operation *op) override;
+
+  std::string outputDir;
+
+  uint32_t counter = 0;
+};
+
+void ReproducerBeforeAll::runBeforePass(Pass *pass, Operation *op) {
+  // Skip adator passes (which adopt FuncOp passes to ModuleOp pass managers).
+  if (isa<OpToOpPassAdaptor>(pass))
+    return;
+
+  llvm::SmallString<128> path(outputDir);
+  if (failed(createDirectoryOrPrintErr(path)))
+    return;
+
+  // Open output file.
+  std::string fileName =
+      llvm::formatv("{0,0+2}_{1}.mlir", counter++, pass->getArgument());
+  llvm::sys::path::append(path, fileName);
+
+  std::string error;
+  std::unique_ptr<llvm::ToolOutputFile> file = openOutputFile(path, &error);
+  if (!file) {
+    llvm::errs() << "Error opening output file " << path << ": " << error
+                 << "\n";
+    return;
+  }
+
+  SmallVector<OperationName> scopes;
+  scopes.push_back(op->getName());
+  while (Operation *parentOp = op->getParentOp()) {
+    scopes.push_back(parentOp->getName());
+    op = parentOp;
+  }
+
+  std::string pipelineStr;
+  llvm::raw_string_ostream passOS(pipelineStr);
+  // Add pass scopes like 'builtin.module(emitc.tu('
+  for (OperationName scope : llvm::reverse(scopes))
+    passOS << scope << "(";
+  pass->printAsTextualPipeline(passOS);
+  for (unsigned i = 0, e = scopes.size(); i < e; ++i)
+    passOS << ")";
```

Comment:
Perhaps split this in a function, it would be a good fit to go next to the parser.

Reply:
I have split it into a function in the same file. I'm not really sure where it would go in the more core parts of MLIR (close to the pipeline parser seems to require new dependencies on other libs).
I will discuss with upstream about the right place when the proposing it there. What do you think?